format/postgres/flavours/postgres14/common14/pgheap.go

```diff
@@ -0,0 +1,498 @@
+
+// typedef uint16 LocationIndex;
+// #define SizeOfPageHeaderData (offsetof(PageHeaderData, pd_linp))
+
+// type = struct HeapTupleHeaderData {
+/*    0      |    12 */ // union {
+/*                12 */ //     HeapTupleFields t_heap;
+/*                12 */ //     DatumTupleFields t_datum;
+//
+/* total size (bytes):   12  */
+/* } t_choice;             //
+/*   12      |     6 */ // ItemPointerData t_ctid;
+/*   18      |     2 */ // uint16 t_infomask2;
+/*   20      |     2 */ // uint16 t_infomask;
+/*   22      |     1 */ // uint8 t_hoff;
+/*   23      |     0 */ // bits8 t_bits[];
+/* XXX  1-byte padding  */
+//
+/* total size (bytes):   24 */
+const SizeOfHeapTupleHeaderData = 24
+
+// type = struct HeapTupleFields {
+/*    0      |     4 */ // TransactionId t_xmin;
+/*    4      |     4 */ // TransactionId t_xmax;
+/*    8      |     4 */ // union {
+/*                 4 */ //    CommandId t_cid;
+/*                 4 */ //    TransactionId t_xvac;
+//                         } t_field3;
+/*                      total size (bytes):    4 */
+//
+/* total size (bytes):   12 */
+
+// type = struct DatumTupleFields {
+/*    0      |     4 */ // int32 datum_len_;
+/*    4      |     4 */ // int32 datum_typmod;
+/*    8      |     4 */ // Oid datum_typeid;
+//
+/* total size (bytes):   12 */
+
+// type = struct ItemPointerData {
+/*    0      |     4 */ // BlockIdData ip_blkid;
+/*    4      |     2 */ // OffsetNumber ip_posid;
+//
+/* total size (bytes):    6 */
+
+type HeapD struct {
+	PageSize uint64
+
+	// current Page
+	Page *HeapPageD
+	// current tuple
+	Tuple *TupleD
+
+	DecodePageHeaderDataFn func(heap *HeapD, d *decode.D)
+	DecodePageSpecialFn    func(heap *HeapD, d *decode.D)
+}
+
+type HeapPageD struct {
+	PdLower           uint16
+	PdUpper           uint16
+	PdSpecial         uint16
+	PdPagesizeVersion uint16
+
+	ItemIds      []itemIDDataD
+	PagePosBegin uint64
+	ItemsEnd     int64
+
+	// pgproee
+	PdXidBase   uint64 // 8 TransactionId pd_xid_base;
+	PdMultiBase uint64 // 8 TransactionId pd_multi_base;
+	PdPruneXid  uint64 // 4 ShortTransactionId pd_prune_xid;
+	PdMagic     uint64 // 4 uint32 pd_magic;
+}
+
+type TupleD struct {
+	IsMulti uint64
+}
+
+type itemIDDataD struct {
+	/*    0: 0   |     4 */ // unsigned int lp_off: 15
+	/*    1: 7   |     4 */ // unsigned int lp_flags: 2
+	/*    2: 1   |     4 */ // unsigned int lp_len: 15
+	lpOff                   uint32
+	lpFlags                 uint32
+	lpLen                   uint32
+}
+
+func DecodeHeap(heap *HeapD, d *decode.D) any {
+	d.SeekAbs(0)
+	decodeHeapPages(heap, d)
+
+	return nil
+}
+
+func decodeHeapPages(heap *HeapD, d *decode.D) {
+	for {
+		if end, _ := d.TryEnd(); end {
+			return
+		}
+
+		page := &HeapPageD{}
+		heap.Page = page
+
+		d.FieldStruct("HeapPage", func(d *decode.D) {
+			decodeHeapPage(heap, d)
+		})
+
+		// end of Page
+		endLen := uint64(d.Pos() / 8)
+		pageEnd := common.TypeAlign(heap.PageSize, endLen)
+		d.SeekAbs(int64(pageEnd) * 8)
```

Comment:
Is it possible that a broken file etc can make this seek back to the same page again causing an infinite loop?